### ISSUE TITLE
S2971 should not be triggered on ToList().AsEnumerable()

### ISSUE THREAD

**Issue by wertzui** [SonarSource/sonar-csharp#1149]
### Description

`var myEnumerable = someEnumerable.ToList().AsEnumerable();` should not trigger S2971 as the AsEnumerable() here is used to cast the result to `IEnumerable<T>`

### Repro steps

Write some code where `ToList()` or `ToArray()` is followed by `.AsEnumerable()`

### Expected behavior

S2971 is not riased.

### Actual behavior

S2971 is raised.

### Known workarounds

None.

### Related information

* SonarAnalyzer.CSharp 6.7.1.4347 NuGet Package
* Visual Studio 2017

**Comment by michalb-sonar** [comment]
@wertzui what is the scenario where you would use `enumerable.ToList().AsEnumerable()`?
Isn't it simpler to write `var myEnumerable = someEnumerable.ToList(<some code>)` (as list is an enumerable)?

**Comment by wertzui** [comment]
This is the real code where the problem occured:
```
var advertisementBlocksByClientId = advertisementBlocks
                .GroupBy(b => b.Inventory.ClientId)
                .ToDictionary(g => g.Key, g => g.ToList().AsEnumerable());
```

So the case is when grouping elements and saving the result in a `Dictionary<TKey, IEnumerable<TValue>>`

**Comment by Evangelink** [comment]
@michalb-sonar I am in favor of implementing this change for the following 2 reasons:
1/ Better readability than cast when used in chained.
2/ (disclaimer: I haven't tested that issue is raised) In the case you have something coming from the database and you do `databaseInfo.Where(something)` the filter will be done using the database but doing `databaseInfo.AsEnumerable().Where(something)` it will be done in memory.

**Comment by michalb-sonar** [comment]
Apologies, forgot about this ticket.

Thanks for the report @wertzui! That is a good use case we haven't thought about.

@Evangelink re 2: good point, I fully expected the call to `AsEnumerable()` be optimized out, but it is still there in IL. Haven't tested with database either, but you may be right.

**Comment by Evangelink** [comment]
I suggest not to raise the rule if `AsEnumerable()` is the last part of the chain. We might end up with some False Negative but they should be relatively rare. WDYT?

**Comment by valhristov** [comment]
We could raise for `AsEnumerable` only if the class has no methods that could hide the Linq extension methods. In other words:

``` C#
var list1 = new List<int>();
list1.AsEnumerable(); // Noncompliant, AsEnumerable() is redundant
var list2 = new MySpecialList(); // MySpecialList declares a Count() method for example
list2.AsEnumerable() // Compliant, you cannot call Linq's Count() method othwerise
```

**Comment by Evangelink** [comment]
That sounds indeed better but might be a lot harder to write.

**Comment by valhristov** [comment]
We could actually check if the class has a method with the same signature as the one that's called immediately after `AsEnumerable`.

**Comment by valhristov** [comment]
Ok, let's exclude `AsEnumerable` entirely.
